File: hackerrank/problem solving/algorithms/2 - implementation/20 - designer pdf viewer/with morfo library/designer_pdf_viewer.java
// https://www.hackerrank.com/challenges/designer-pdf-viewer/problem?isFullScreen=true
// Java 25

import morfo.io.Reader;

void main() {
    int[] lettersHeights = Reader.read(26, Integer::parseInt).stream().mapToInt(Integer::intValue).toArray();
    String word = Reader.read(s -> s);
    IO.println(designerPdfViewer(lettersHeights, word));
}

// n: length word, <= 10
// T: O(10) = O(1)
// S: O(1) extra space
int designerPdfViewer(int[] lettersHeights, String word) {
    return word.chars().map(ch -> lettersHeights[ch - 'a']).max().orElse(0) * word.length();
}

File: hackerrank/problem solving/algorithms/2 - implementation/20 - designer pdf viewer/with morfo library/designer_pdf_viewer.cs
// https://www.hackerrank.com/challenges/designer-pdf-viewer/problem?isFullScreen=true

using Morfo.IO;

int[] lettersHeights = Reader.Read<int>(26);
string word = Reader.Read<string>();
Console.WriteLine(DesignerPdfViewer(lettersHeights, word));

// n: length word, <= 10
// T: O(10) = O(1)
// S: O(1) extra space
int DesignerPdfViewer(int[] lettersHeights, string word)
{
    return word.Max(ch => lettersHeights[ch - 'a']) * word.Length;
}
